cellprofiler/modules/classify.py

```diff
@@ -0,0 +1,952 @@
+        :param sampling: a sampling of the ground truth
+        :param classes: a vector of the same length as the sampling giving the indexes of the classes of each sample
+        :param filter_bank_name: the name to assign to the filter bank
+        :param n_filters: # of filters to create
+        :param algorithm: an object that can be fitted using algorithm.fit(X, Y) and can transform using algorithm.transform(X). Default is RandomizedPCA.
+
+        :return:
+        """
+        if algorithm is None:
+            algorithm = sklearn.decomposition.RandomizedPCA(n_filters, random_state=self.random_state(filter_bank_name))
+
+        algorithm.fit(sampling, classes)
+
+        if hasattr(algorithm, "components_"):
+            components = algorithm.components_
+
+            if len(components) > n_filters:
+                components = components[:n_filters]
+
+            if filter_bank_name in self.g_filters.keys():
+                del self.g_filters[filter_bank_name]
+
+            self.g_filters.create_dataset(filter_bank_name, data=components).attrs[A_CLASS] = CLASS_FILTER
+        else:
+            self.g_filters.create_dataset(filter_bank_name, data=pickle.dumps(algorithm)).attrs[A_CLASS] = CLASS_CLASSIFIER
+
+    def use_filter_bank(self, filter_bank_name, sample):
+        """
+        Transform a sample using a filter bank
+
+        :param filter_bank_name:
+        :param sample:
+
+        :return:
+        """
+        dataset = self.g_filters[filter_bank_name]
+
+        if dataset.attrs[A_CLASS] == CLASS_FILTER:
+            if USE_DOT:
+                result = numpy.dot(sample, dataset[:].T)
+            else:
+                #
+                # A dot product... but cluster's np.dot is so XXXXed
+                #
+                chunk_size = self.ground_truth_chunk_size
+
+                result = []
+
+                for index in range(0, len(sample), chunk_size):
+                    index_end = min(index + chunk_size, len(sample))
+
+                    logger.debug("Processing dot product chunk {0:d}:{1:d} of {2:d}".format(index, index_end, len(sample)))
+
+                    result.append(numpy.sum(sample[index:index_end, :, numpy.newaxis] * dataset[:].T[numpy.newaxis, :, :], 1))
+
+                result = numpy.vstack(result)
+        else:
+            algorithm = pickle.loads(dataset.value)
+
+            result = algorithm.transform(sample)
+
+        return result
+
+    def fit(self, classifier_name, sample, classes, algorithm=None):
+        """
+        Fit an algorithm to data and save
+
+        :param classifier_name: save using this name
+        :param sample: S samples x N features
+        :param classes: S class labels indexing into the class names
+        :param algorithm: algorithm to use to train
+
+        :return:
+        """
+        if algorithm is None:
+            algorithm = sklearn.ensemble.ExtraTreesClassifier(n_estimators=N_ESTIMATORS, min_samples_leaf=MIN_SAMPLES_PER_LEAF)
```

Comment:
Good catch on N_ESTIMATORS and MIN_SAMPLES_PER_LEAF. That's a legacy of me fooling around with the classifier when those were static variables.